src/lib/comms/sol-http-common.c

```diff
@@ -882,3 +883,238 @@ sol_http_split_post_field(const char *query, struct sol_http_params *params)
 {
     return sol_http_split_str_key_value(query, SOL_HTTP_PARAM_POST_FIELD, params);
 }
+
+static int
+sort_priority(const void *v1, const void *v2)
+{
+    const struct sol_http_content_type_priority *pri1, *pri2;
+    bool any_type1, any_type2;
+
+    pri1 = (const struct sol_http_content_type_priority *)v1;
+    pri2 = (const struct sol_http_content_type_priority *)v2;
+
+    any_type1 = sol_str_slice_str_eq(pri1->type, "*");
+    any_type2 = sol_str_slice_str_eq(pri2->type, "*");
+
+    //text/html have precedence over */* for example.
+    if (!any_type1 && any_type2)
+        return -1;
+    if (any_type1 && !any_type2)
+        return 1;
+
+    //Higher qvalue first
+    if (pri1->qvalue > pri2->qvalue)
+        return -1;
+    if (pri1->qvalue < pri2->qvalue)
+        return 1;
+
+    //Specialized first
+    if (pri1->tokens.len > pri2->tokens.len)
+        return -1;
+    if (pri1->tokens.len < pri2->tokens.len)
+        return 1;
+
+    //Specialized first
+    if (sol_str_slice_eq(pri1->type, pri2->type)) {
+        //text/html have precedence over text/* for example.
+        any_type1 = sol_str_slice_str_eq(pri1->sub_type, "*");
+        any_type2 = sol_str_slice_str_eq(pri2->sub_type, "*");
+
+        if (!any_type1 && any_type2)
+            return -1;
+        if (any_type1 && !any_type2)
+            return 1;
+    }
+
+    //Precedence must prevail
+    if (pri1->index > pri2->index)
+        return 1;
+    if (pri1->index < pri2->index)
+        return -1;
+
+    return 0;
+}
+
+static char *
+is_qvalue_param(const struct sol_str_slice param)
+{
+#define STATE_NEEDS_Q_CHAR (0)
+#define STATE_NEEDS_EQUAL_CHAR (1)
+
+    size_t i;
+    int state = STATE_NEEDS_Q_CHAR;
+
+    for (i = 0; i < param.len; i++) {
+        if (isspace((int)param.data[i]))
+            continue;
+
+        switch (state) {
+        case STATE_NEEDS_Q_CHAR:
+            if (param.data[i] != 'q')
+                return NULL;
+            state = STATE_NEEDS_EQUAL_CHAR;
+            break;
+        default:
+            if (param.data[i] != '=')
+                return NULL;
+            return (char *)param.data + i;
```

Comment:
it's better to return the qvalue_slice, you'll need it anyway. Return an empty slice if not found.